paasta_tools/kubernetes/bin/paasta_secrets_sync.py

```diff
@@ -140,34 +144,66 @@ def main() -> None:
     ) else sys.exit(1)
 
 
-def get_services_to_k8s_namespaces(
+def get_services_to_k8s_namespaces_to_allowlist(
     service_list: List[str], cluster: str, soa_dir: str, kube_client: KubeClient
-) -> Dict[str, Set[str]]:
-    services_to_k8s_namespaces: Dict[str, Set[str]] = defaultdict(set)
+) -> Tuple[
+    Dict[
+        str,  # service
+        Dict[
+            str,  # namespace
+            Optional[Set[str]],  # allowlist of secret names, None means allow all.
+        ],
+    ],
+]:
+    """
+    Generate a mapping of service -> namespace -> allowlist of secrets, e.g.
+
+    {
+        "yelp-main": {
+            "paasta": {"secret1", "secret2"},
+            "paastasvc-yelp-main": {"secret1", "secret3"},
+            "paasta-flinks": None,
+        },
+        "_shared": {
+            "paasta": {"sharedsecret1"},
+            "paastasvc-yelp-main": {"sharedsecret1", "sharedsecret2"},
+            "paasta-flinks": None,
```

Comment:
i wonder if this is a good opportunity to define `set()` as meaning there should be no secrets to allow us to truly sync things (i.e., lay the groundwork for deleting secrets that are no longer used rather than leaving expired keys in the k8s secret store forever)

Reply:
With the way the code is written, `set()` does mean nothing is allowed, which is why I used `None` for allow everything.

Reply:
yea, or more completely: set() for a given service/namespace means we should delete all secrets that had previously been synced for that service in that namespace. 

a second step would then be to actually take the difference between the synced set and allowlist set and delete that per-service/namespace

(although, it's probably easier to do this all in one go at the end of syncing for shared namespaces like paasta-spark as you can union all the secret names that should exist and the delete the difference between the previously synced set and the union of allowlist sets)